sdk/go/docker/docker.go

```diff
@@ -0,0 +1,558 @@
+// Copyright 2016-2020, Pulumi Corporation.
+
+package docker
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math/rand"
+	"os/exec"
+	"strings"
+
+	"github.com/Masterminds/semver"
+
+	"github.com/pkg/errors"
+	"github.com/pulumi/pulumi/pkg/util/logging"
+	"github.com/pulumi/pulumi/sdk/go/pulumi"
+)
+
+func buildAndPushImageAsync(ctx *pulumi.Context, imageName string, build DockerBuild, repositoryURL string,
+	logResource pulumi.Resource, skipPush bool, registry ImageRegistry) (string, error) {
+
+	// Give an initial message indicating what we're about to do.  That way, if anything
+	// takes a while, the user has an idea about what's going on.
+	logging.Infof("Starting docker build and push...")
+
+	res, err := buildAndPushImageWorkerAsync(ctx, imageName, build, repositoryURL, logResource, skipPush, &registry)
+	if err != nil {
+		return "", err
+	}
+
+	// If we got here, then building/pushing didn't throw any errors.  update the status bar
+	// indicating that things worked properly.  that way, the info bar isn't stuck showing the very
+	// last thing printed by some subcommand we launched.
+	logging.Infof("Successfully pushed to docker.")
+	return res, nil
+}
+
+func buildAndPushImageWorkerAsync(ctx *pulumi.Context, baseImageName string, build DockerBuild,
+	repositoryURL string, logResource pulumi.Resource, skipPush bool, registry *ImageRegistry) (string, error) {
+
+	err := checkRepositoryURL(repositoryURL)
+	if err != nil {
+		return "", err
+	}
+
+	_, tag := getImageNameAndTag(baseImageName)
+
+	// login immediately if we're going to have to actually communicate with a remote registry.
+	//
+	// We know we have to login if:
+	//
+	//  1. We're doing an update.  In that case, we'll always want to login so we can push our
+	//     images to the remote registry.
+	//
+	// 2. We're in preview or update and the build information contains 'cache from' information. In
+	//    that case, we'll want want to pull from the registry and will need to login for that.
+	pullFromCache := build.CacheFrom != nil && len(repositoryURL) > 0
+
+	// If no `registry` info was passed in we simply assume docker is already
+	// logged-in to the correct registry (or uses auto-login via credential helpers).
+	if registry != nil {
+		if !ctx.DryRun() || pullFromCache {
+			logging.Infof("Logging into registry...")
+			err := loginToRegistry(*registry, logResource)
+			if err != nil {
+				return "", err
+			}
+		}
+	}
+
+	// If the container specified a cacheFrom parameter, first set up the cached stages.
+	var cacheFrom []string
+	if pullFromCache {
+		cacheFrom = pullCacheAsync(baseImageName, *build.CacheFrom, repositoryURL, logResource)
```

Comment:
Here and elsewhere as well - the `Async` suffix can be dropped as none of these are Async in Go.